### PR TITLE
fix Issue 15607 - [ICE] CTFE internal error: bad compare on accessing default-initialized static immutable array of array

### PR DESCRIPTION
**Pull request by WalterBright**
https://issues.dlang.org/show_bug.cgi?id=15607

### CODE REVIEW

**Comment by 9rnsr** [comment]
How do you think about this fix: is it a workaround or a conclusive fix?

Currently, we have a known inconsistency - whether block assignment is accepted, depends on the location a variable is declaring.

```d
void main()
{
    // function scope
    int[][3] a = [1,2,3,4,5];   // OK, block assignment
    int[3][] b = [1,2,3];       // NG, not a block assignment
}

// module scope
int[][3] a = [1,2,3,4,5];   // cannot compile
int[3][] b = [1,2,3];       // cannot compile
```

OT: I've opened #3680 to make array initializer behavior consistent between function scopes and module scopes. It has passed auto-tester, but couldn't get well reviewing about language behavior change. So it's not maintained today.

How do you think about the inconsistency? By accepting the issue 15607 code, are you intending to make all block assignments in module scopes valid in D?

**Comment by WalterBright** [comment]
Array initializers should structurally match the type. I don't know what you mean by 'block assignment'.
```
int[][3] a = [1,2,3,4,5];
int[3][] b = [1,2,3];
```
The first tries to assign 5 elements to a 3 element array, so should not compile. The second is creating a dynamic array of 3 elements. Each of those elements is used to initialize a 3 element array, so should compile.
These should hold whether at module or local scope.

**Comment by 9rnsr** [comment]
Dmd doesn't work as you thought, for variable initializing in function scope.

"Block assignment" is: a static array type variable can be assigned by its element type value, because of the implicit slicing. e.g.:

```d
void foo() {
    int[3] a;
    a = 1;   // block assignment, this is implicitly rewritten to a[] = 1;
    int[3] b = 1;  // also OK, called "block assignment" or "block initializing" in dmd code.
}
```

Note that there is also "multidimensional block initializing" feature.

```d
void foo() {
    int[3][2] a = 1;   // OK, a is initialized to [[1,1,1],[1,1,1]]
    a = 1;  // NG, because a[] = 1; is not correct.
}
```

The feature is caused from the implementation in `AssignExp.semantic`. Therefore, it doesn't work in module scope.

In short, current dmd has two kinds variable initialization behavior for function scope and module scope. If the inconsistency is intentional, is there a design reason to justify it? Otherwise, isn't the inconsistency a dm bug?

**Comment by 9rnsr** [comment]
(sorry I accidentally pushed close button.)

**Comment by WalterBright** [comment]
>  a = 1;   // block assignment, this is implicitly rewritten to a[ ] = 1;

I can't justify that behavior in the compiler. Initialization isn't the same thing as assignment.

**Comment by 9rnsr** [comment]
> Initialization isn't the same thing as assignment.

I know that, and it's not point here.

In `VarDeclaration.semantic`, a variable initialization in local scope is rewritten to `BlitExp` or `ConstructExp`, these are derived classes of `AssignExp`. And the `AssignExp.semantic` with `op != TOKassign` behavior is actually different from `ArrayInitialization.semantic`. I'm talking about the real difference. Is it a bug or feature?

**Comment by WalterBright** [comment]
> I can't justify that behavior in the compiler

means it's a bug

**Comment by 9rnsr** [comment]
> means it's a bug

Are you sure? Today, following block initializations in local scope are accepted by dmd.

```d
void testBlockInitialization()
{
    // The elements of static array int[2]
    // are initialized by an int element value 1.
    int[2] a1 = 1;

    // The elements of multidimensional static array int[2][2]
    // are initialized by an int element value 1.
    int[2][2] a2 = 1;

    // The elements of static array int[][2]
    // are initialized by an int[] element value [1,2,3].
    int[][3] a3 = [1,2,3];

    // The elements of multidimensional static array int[][2][2]
    // are initialized by an int[] element value [1,2,3].
    int[][2][2] a4 = [1,2,3];

    import std.stdio;
    writeln(a1);    // [1, 1]
    writeln(a2);    // [[1, 1], [1, 1]]
    writeln(a3);    // [[1, 2, 3], [1, 2, 3], [1, 2, 3]]
    writeln(a4);    // [[[1, 2, 3], [1, 2, 3]], [[1, 2, 3], [1, 2, 3]]]
}
```

Are you saying they should become invalid?

**Comment by WalterBright** [comment]
> following block initializations in local scope are accepted by dmd.

These are structurally different from the earlier examples you gave.

> Dmd doesn't work as you thought

I can't catch everything going wrong in the thousands of PRs.

The bottom line is something is terribly wrong when local array initializers are different from module initializers.

**Comment by 9rnsr** [comment]
> I can't catch everything going wrong in the thousands of PRs.

> The bottom line is something is terribly wrong when local array initializers are different from module initializers.

The issue has existed since 2010, when I've started to contribute to dmd by using svn.

**Comment by WalterBright** [comment]
Bringing up these other issues deserves a separate thread, since they are not relevant to this fix.

File: src/init.d [hunk 1]
```diff
@@ -692,12 +694,41 @@ public:
...
+            if (t)
+            {
+                Type tn = t.nextOf().toBasetype();
+                if (tn.ty == Tsarray)
+                {
+                    const dim = cast(size_t)(cast(TypeSArray)tn).dim.toInteger();
+                    Type te = tn.nextOf().toBasetype();
+                    foreach (ref e; *elements)
+                    {
+                        if (te.equals(e.type))
```
Comment (andralex) [inline src/init.d]: Use the "exhaustive `if`" idiom:

```D
foreach (...)
{
    if (...)
    {
        ... code 1 ...
        ... code 2 ...
        ... code 3 ...
        ... code 4 ...
        ... code 5 ...
        ... code 6 ...
    }
}
```
=====>>>>
```D
foreach (...)
{
    if (...)
        continue;
    ... code 1 ...
    ... code 2 ...
    ... code 3 ...
    ... code 4 ...
    ... code 5 ...
    ... code 6 ...
}
```

**Comment by andralex** [comment]
Auto-merge toggled on
